drivers/adc/adc_ad405x.c

```diff
@@ -0,0 +1,928 @@
+#define AD405X_REG_ADC_MODES          0x21U
+#define AD405X_REG_AVG_CONFIG         0x23U
+#define AD405X_REG_GP_PIN_CONF        0x24U
+#define AD405X_REG_TIMER_CONFIG       0x27U
+
+#define AD405X_REG_INTERFACE_CONFIG_A_VAL       0x10U
+#define AD405X_REG_DEVICE_TYPE_VAL              0x07U
+#define AD4052_REG_PRODUCT_ID_VAL               0x0072U
+#define AD4050_REG_PRODUCT_ID_VAL               0x0070U
+#define AD405X_REG_VENDOR_VAL                   0x0456U
+#define AD405X_REG_INTERFACE_CONFIG_A_RESET_VAL 0x81U
+
+/** AD405X_REG_ADC_MODES Bit definition */
+#define AD405X_ADC_MODES_MSK        GENMASK(2, 0)
+#define AD405X_BURST_AVERAGING_MODE BIT(0)
+#define AD405X_AVERAGING_MODE       BIT(1)
+
+/** AD405X_REG_MODE_SET Bit definition */
+#define AD405X_ENTER_ADC_MODE_MSK BIT(0)
+#define AD405X_ENTER_ADC_MODE     BIT(0)
+#define AD405X_ENTER_SLEEP_MODE   BIT(1) | BIT(0)
+#define AD405X_ENTER_ACTIVE_MODE  0x0U
+
+/** AD405X_REG_AVG_CONFIG Bit Definitions */
+#define AD405X_AVG_WIN_LEN_MSK GENMASK(3, 0)
+
+#define AD405X_WRITE_CMD 0x0U
+#define AD405X_READ_CMD  0x80U
+
+#define AD405X_SW_RESET_MSK   BIT(7) | BIT(0)
+
+#define AD405X_GP1_MODE_MSK GENMASK(6, 4)
+#define AD405X_GP0_MODE_MSK GENMASK(2, 0)
+#define AD405X_GP1          0x1U
+#define AD405X_GP0          0x0U
+
+/** AD405X_REG_TIMER_CONFIG Bit Definitions */
+#define AD405X_FS_BURST_AUTO_MSK GENMASK(7, 4)
+
+union ad405x_bus {
+	struct spi_dt_spec spi;
+};
+
+enum ad405x_gpx_mode {
+	AD405X_DISABLED = 0,
+	AD405X_GP0_1_INTR = 1,
+	AD405X_DATA_READY = 2,
+	AD405X_DEV_ENABLE = 3,
+	AD405X_CHOP = 4,
+	AD405X_LOGIC_LOW = 5,
+	AD405X_LOGIC_HIGH = 6,
+	AD405X_DEV_READY = 7
+};
+
+/** AD405X modes of operations */
+enum ad405x_operation_mode {
+	AD405X_SAMPLE_MODE_OP = 0,
+	AD405X_BURST_AVERAGING_MODE_OP = 1,
+	AD405X_AVERAGING_MODE_OP = 2,
+	AD405X_MONITOR_AUTO_MODE_OP = 3,
+	AD405X_CONFIG_MODE_OP = 4,
+	AD405X_SLEEP_MODE_OP = 5,
+	AD405X_TRIGGER_AUTO_MODE_OP = 7
+};
+
+/** AD405X sample rate for burst and autonomous modes */
+enum ad405x_sample_rate {
+	AD405X_2_MSPS,
+	AD405X_1_MSPS,
+	AD405X_333_KSPS,
+	AD405X_100_KSPS,
+	AD405X_33_KSPS,
+	AD405X_10_KSPS,
+	AD405X_3_KSPS,
+	AD405X_1_KSPS,
+	AD405X_500_SPS,
+	AD405X_333_SPS,
+	AD405X_250_SPS,
+	AD405X_200_SPS,
+	AD405X_166_SPS,
+	AD405X_140_SPS,
+	AD405X_125_SPS,
+	AD405X_111_SPS
+};
+
+/** AD405X averaging filter window length */
+enum ad405x_avg_filter_l {
+	AD405X_LENGTH_2,
+	AD405X_LENGTH_4,
+	AD405X_LENGTH_8,
+	AD405X_LENGTH_16,
+	AD405X_LENGTH_32,
+	AD405X_LENGTH_64,
+	AD405X_LENGTH_128,
+	AD405X_LENGTH_256,
+	AD405X_LENGTH_512,
+	AD405X_LENGTH_1024,
+	AD405X_LENGTH_2048,
+	AD405X_LENGTH_4096
+};
+static const int avg_filter_values[] = {2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096};
+
+struct adc_ad405x_config {
+	const union ad405x_bus bus;
+	enum ad405x_operation_mode active_mode;
+#ifdef CONFIG_AD405X_TRIGGER
+	struct gpio_dt_spec gp1_interrupt;
+	struct gpio_dt_spec gp0_interrupt;
+	uint8_t has_gp1;
+	uint8_t has_gp0;
```

Comment:
These can be replaced by checking if the port is NULL in `gp1_interrupt` and `gp0_interrupt`
```
uint8_t has_gp1 = config->gp1_interrupt->port != NULL;
```